widget/toolbar.go

```diff
@@ -13,31 +13,48 @@ type ToolbarItem interface {
 	ToolbarObject() fyne.CanvasObject
 }
 
+type toolbarActionButton struct {
+	Button
+
+	t *ToolbarAction
+}
+
+func (b *toolbarActionButton) Refresh() {
+	b.Icon = b.t.Icon
+	b.OnTapped = b.t.OnActivated
+
+	if b.t.Disabled {
+		b.Button.Disable()
+	} else {
+		b.Button.Enable()
+	}
+}
+
 // ToolbarAction is push button style of ToolbarItem
 type ToolbarAction struct {
 	Icon        fyne.Resource
 	OnActivated func() `json:"-"`
+	// Since: 2.5
+	Disabled bool
 }
 
 // ToolbarObject gets a button to render this ToolbarAction
 func (t *ToolbarAction) ToolbarObject() fyne.CanvasObject {
-	button := NewButtonWithIcon("", t.Icon, t.OnActivated)
-	button.Importance = LowImportance
-
-	return button
+	b := &toolbarActionButton{t: t}
+	b.Refresh()
```

Comment:
Is the Refresh here necessary? I don't think so since this function should only be called to get a new Button to render the toolbar action, which is immediately placed on the canvas for display which will cause it to get refreshed automatically?

Reply:
The toolbarRenderer `Refresh()` code does not call the `Refresh()` function of the underlying objects, but only creates new ones using `ToolbarObject()`. Since the underlying `Button` of the `toolbarActionButton` is only updated with the `Refresh()` call, which is not happening by the `toolbarRenderer`, I chose to implement it here. Alternatively, I can also add a `Refresh()` call in the `toolbarRenderer`:

https://github.com/fyne-io/fyne/blob/ccc6a1c19a6d16988da3e468f08f25d8100cd4b4/widget/toolbar.go#L145

Let me know what you prefer.

Reply:
Typically widgets are created with the correct state so that a Refresh is not required before they first draw.

> The toolbarRenderer Refresh() code does not call the Refresh() function of the underlying objects, but only creates new ones using ToolbarObject()

Given this information, and that as far as I can see the only way to ask this item to refresh is to call `ToolBar.Refresh()` once the item is `Disabled`, is the custom type and additional refresh paths needed?

Reply:
This call to refresh would not be needed if the disabled state is set correctly when you create the button - that would match how the code works elsewhere.

Reply:
I think this would only move the call to `Refresh` into the constructor of the `toolbarActionButton` as I also need to make sure that the `Icon` & the `OnTapped` handler are set correctly.

Reply:
What I am thinking is no refresh should be needed here. Setting the disabled state on creation would avoid it would it not?

Reply:
Are you thinking about something like this?

```
func newToolbarActionButton(t *ToolbarAction) *toolbarActionButton {
	b := &toolbarActionButton{t: t}
	b.ExtendBaseWidget(b)

	b.Icon = t.Icon
	b.OnTapped = b.t.OnActivated
	b.disabled = b.t.Disabled

	return b
}
```

Note that I also need to set the `Icon` & `OnTapped` handler here, since we are not calling `Refresh` afterwards, and therefore these fields would not be set.

Having this code though results in a nasty duplication in my opinion, because the same fields are set when the state is synchronized with the `ToolbarAction`

```
func newToolbarActionButton(t *ToolbarAction) *toolbarActionButton {
	b := &toolbarActionButton{t: t}
	b.ExtendBaseWidget(b)

	b.Icon = t.Icon
	b.OnTapped = b.t.OnActivated
	b.disabled = b.t.Disabled

	return b
}

func (b *toolbarActionButton) Refresh() {
	b.Icon = b.t.Icon
	b.OnTapped = b.t.OnActivated

	if b.t.Disabled {
		b.Button.Disable()
	} else {
		b.Button.Enable()
	}
}
```

I would propose to wait for the results of our discussion in https://github.com/fyne-io/fyne/pull/4347#discussion_r1439335003, as this could change this affected code part.